src/models/marian.cpp

```diff
@@ -0,0 +1,244 @@
+
+    *past_key_values_length_->GetTensorMutableData<int64_t>() = -1;
+
+    attention_mask_.name_ = model_.config_->model.decoder.inputs.encoder_attention_mask;
+    attention_mask_.Add();
+    attention_mask_.Update(next_tokens, current_length, static_cast<int>(new_length));
+
+    auto hidden_states_type = model_.session_info_.GetInputDataType("encoder_hidden_states");
+    auto encoder_hidden_states_shape = std::array<int64_t, 3>{decoder_input_ids_.GetMarianInputsShape()[0], encoder_input_ids_.GetShape()[1], 512};
+    encoder_hidden_states_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), encoder_hidden_states_shape, hidden_states_type);
+
+    input_names_.push_back("encoder_hidden_states");
+    inputs_.push_back(encoder_outputs_.get());
+
+    auto rnn_states_prev_type = model_.session_info_.GetInputDataType("rnn_states_prev");
+    auto rnn_states_prev_shape = std::array<int64_t, 3>{3, decoder_input_ids_.GetMarianInputsShape()[0], 512};
+    rnn_states_prev_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), rnn_states_prev_shape, rnn_states_prev_type);
+
+    input_names_.push_back("rnn_states_prev");
+    for (int i = 0; i < rnn_states_prev_shape[0]; i++) {
+      for (int j = 0; j < rnn_states_prev_shape[1]; j++) {
+        for (int k = 0; k < rnn_states_prev_shape[2]; k++) {
+          auto data = rnn_states_prev_->GetTensorMutableData<int32_t>();
+          data[i * rnn_states_prev_shape[1] * rnn_states_prev_shape[2] + j * rnn_states_prev_shape[2] + k] = 0;
+        }
+      }
+    }
+    inputs_.push_back(rnn_states_prev_.get());
+
+    auto rnn_states_type = model_.session_info_.GetOutputDataType("rnn_states");
+    auto rnn_states_shape = std::array<int64_t, 3>{3, decoder_input_ids_.GetMarianInputsShape()[0], 512};
+    rnn_states_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), rnn_states_shape, rnn_states_type);
+
+    output_names_.push_back("rnn_states");
+    outputs_.push_back(rnn_states_.get());
+    *past_key_values_length_->GetTensorMutableData<int64_t>() += 1;
+
+    logits_.Add();
+
+    logits_.Update(next_tokens.subspan(next_tokens.size() - 1, 1), 1);
+
+    State::Run(*model_.session_decoder_);
+    first_run_ = false;
+    return logits_.Get();
+  }
+
+  // UPDATE THE DECODER
+  decoder_input_ids_.Update(next_tokens);
+  auto rnn_states_prev_shape = std::array<int64_t, 3>{3, decoder_input_ids_.GetMarianInputsShape()[0], 512};
+
+  for (int i = 0; i < rnn_states_prev_shape[0]; i++) {
+    for (int j = 0; j < rnn_states_prev_shape[1]; j++) {
+      for (int k = 0; k < rnn_states_prev_shape[2]; k++) {
+        auto data = rnn_states_prev_->GetTensorMutableData<int32_t>();
+        auto rnn_states_data = rnn_states_->GetTensorMutableData<int32_t>();
+        data[i * rnn_states_prev_shape[1] * rnn_states_prev_shape[2] + j * rnn_states_prev_shape[2] + k] = rnn_states_data[i * rnn_states_prev_shape[1] * rnn_states_prev_shape[2] + j * rnn_states_prev_shape[2] + k];
+      }
+    }
+  }
```

Comment:
Avoid accessing the device allocated input/output through CPU code. You can use the device span interface for this.